blog/_posts/2021-11-22-k8s-custom-deploy.md

```diff
@@ -0,0 +1,123 @@
+It's also common for teams to have their own custom deployment shell scripts (e.g. for use with CI/CD).
+
+While it's possible to use Tilt's [`local_resource`][api-local_resource] function as a workaround in these instances, that means giving up much of Tilt's deep Kubernetes integration for the resource as a result.
+
+In Tilt [v0.23.0][tilt-releases], we've introduced a built-in function to enable the use of an external Kubernetes deployment tool or script.
+With the new [`k8s_custom_deploy`][api-k8s_custom_deploy] function, Tilt will delegate to a command or shell script to perform the deployment instead of directly calling the Kubernetes API with YAML.
+Tilt features like status reporting and log streaming work automatically.
+Even the more advanced customizations provided by [`k8s_resource`][api-k8s_resource] such as port forwards are supported!
+
+## Example
+Let's use [`k8s_custom_deploy`][api-k8s_custom_deploy] to deploy an [OpenFaaS][openfaas] function without writing a Dockerfile or Kubernetes YAML.
+
+For this example, I've installed the OpenFaaS components in my cluster manually; in a real Tiltfile, you could include this as well to ensure the local environment is set up properly.
+The [OpenFaaS blog][openfaas-blog-tilt] has a guide that might be useful.
+
+I'm using a Golang handler bootstrapped by running `faas-cli new go-fn --lang go`.
+This created a `go-fn.yml` file (this is an OpenFaaS config, NOT a Kubernetes manifest.)
+It also made a `go-fn/` directory with "Hello world" example handler code.
+
+```python
+# on `tilt up` and file changes -> run the OpenFaaS build + deploy and then query for what it deployed
+apply_cmd = """
+faas-cli up -f go-fn.yml 1>&2
+kubectl get -oyaml --namespace=openfaas-fn all -l "faas_function=go-fn"
+"""
+
+# on `tilt down` -> delete the OpenFaaS function from the cluster
+delete_cmd = 'faas-cli delete -f go-fn.yml'
+
+k8s_custom_deploy(
+    'go-fn',
+    apply_cmd=apply_cmd,
+    delete_cmd=delete_cmd,
+    # apply_cmd will be re-executed whenever these files change
+    deps=['./go-fn.yml', './go-fn/']
+)
+
+# add a port forward so we can debug the function from our host machine directly
+k8s_resource('go-fn', port_forwards=['9372:8080'])
+```
+
+When we run `tilt up`, Tilt will execute the `apply_cmd`, which in our case invokes `faas-cli up` to build and deploy an image and then returns the _result_ as YAML so Tilt can track the new or updated Kubernetes objects.
+If `go-fn.yml` or any file in the `go-fn/` directory tree changes, the `apply_cmd` will be automatically re-run.
+
+On `tilt down`, our `delete_cmd` will be invoked, which allows `faas-cli` to delete any objects from Kubernetes it created as well as clean up any extra OpenFaaS state.
+
+## What if Something Goes Wrong?
+Integrating an external tool can be tricky!
+Luckily, the Tilt API allows us to quickly introspect what's happening behind the scenes.
+
+> 💡 Use `tilt describe kapp` to get a human readable version!
```

Comment:
IMO this should go after the code snippet because before it, I don't know what you're talking about.